### PR TITLE
add detect_sharp_corners

### LINKED ISSUE #1708: Improvements for PMP::isotropic_remeshing
We should 

- [ ] deal with the fact that `Patch_id_pmap` may not have been initialized in the Polyhedron demo
- [ ] add vertices attached to constraints forming a sharp angle as constrained vertices
- [ ] generalize the smoothing iteration of isotropic remeshing: integrate the current smoothing used by istropic remeshing into PMP::smooth_mesh and conversely, use PMP::smooth_mesh to improve the quality of the results of isotropic remeshing (copied from @sloriot comment below)

### PR DESCRIPTION
**Pull request by kabirkedia**
## Summary of Changes

Adds the function detect_sharp_corners to detect sharp corners as addressed in issue #1708 

## Release Management

* Affected package(s):PMP
* Issue(s) solved (if any): fix #1708 
* Feature/Small Feature (if any): detect sharp corners
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership:

### CODE REVIEW

**Comment by janetournois** [comment]
You should add a call to `detect_sharp_corners()` in the example `detect_features_example.cpp` to test it

File: Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp [hunk 1]
```diff
@@ -27,15 +27,25 @@ int main(int argc, char* argv[])
   typedef boost::property_map<Mesh, CGAL::edge_is_feature_t>::type EIFMap;
   typedef boost::property_map<Mesh, CGAL::face_patch_id_t<int> >::type PIMap;
   typedef boost::property_map<Mesh, CGAL::vertex_incident_patches_t<int> >::type VIMap;
+  typedef boost::property_map<Mesh, CGAL::vertex_is_constrained_map>::type VCMap
 
   EIFMap eif = get(CGAL::edge_is_feature, mesh);
   PIMap pid = get(CGAL::face_patch_id_t<int>(), mesh);
   VIMap vip = get(CGAL::vertex_incident_patches_t<int>(), mesh);
+  VCMap vcmap =get(CGAL::vertex_is_constrained_map,meshj);
```
Comment (janetournois) [inline Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp:35]: Does this line compile?

Comment (kabirkedia) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp]: I made an error. I fixed it just now!

Comment (afabri) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp]: And please add the whitespace right after the `=` and the `,` just as it is done in the line above.    This is not done everywhere, but adapt to the local style of a file.

Comment (afabri) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp]: > 
> Does this line compile?

So the general rule (which I unfortunately also too often do not respect) is to compile/run the testsuite of the package you modify, before a git push.

Comment (kabirkedia) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp]: Okay, I'll compile it, fix all the errors and get back!

**Comment by janetournois** [comment]
The example cpp does not compile as is. You should add a version of `detect_sharp_corners()` without the named parameters (similarly to what is done for `detect_sharp_edges()`, using `parameters::all_default()`).

Can you please compile and test the code locally before pushing?

File: Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp [hunk 2]
```diff
@@ -27,15 +27,25 @@ int main(int argc, char* argv[])
   typedef boost::property_map<Mesh, CGAL::edge_is_feature_t>::type EIFMap;
   typedef boost::property_map<Mesh, CGAL::face_patch_id_t<int> >::type PIMap;
   typedef boost::property_map<Mesh, CGAL::vertex_incident_patches_t<int> >::type VIMap;
+  typedef boost::property_map<Mesh, CGAL::vertex_feature_degree_t>::type VIFMap;
```
Comment (janetournois) [inline Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp:30]: The property map you are filling is not the vertex feature degree.
You should create a new one, similar to `edge_is_feature_t` (line 27), that describes what you are listing/detecting. I would suggest `CGAL::vertex_is_feature_t`

Comment (kabirkedia) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp:30]: It's giving an error showing no such thing as vertex_is_feature_t in CGAL.

**Comment by kabirkedia** [comment]
> The example cpp does not compile as is. You should add a version of `detect_sharp_corners()` without the named parameters (similarly to what is done for `detect_sharp_edges()`, using `parameters::all_default()`).
> 
> Can you please compile and test the code locally before pushing?

I did that and resolved all the conflicts except function not found!

File: Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp (hunk 2 above)
Comment (janetournois) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp:30]: right
the parameter and type should be added as `edge_is_feature_t` was previously added

Comment (kabirkedia) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp:30]: How should I add it?

Comment (kabirkedia) [reply Polygon_mesh_processing/examples/Polygon_mesh_processing/detect_features_example.cpp:30]: I added that in the boost/graph/properties.h right now!

**Comment by kabirkedia** [comment]
I have built the file successfully. Can you provide me with some input and suitable output to verify if I've implemented it correctly?

**Comment by janetournois** [comment]
Can you please add a call to `detect_sharp_edges()` function in the cpp example, so that we can test the code?

When it works, it would also be great if you could add another action/submenu to the demo, in the file  `Polyhedron\demo\Polyhedron\Plugins\PMP\Detect_sharp_edges_plugin.cpp`
It is the easiest way to visually test a new function and debug it

**Comment by kabirkedia** [comment]
> Can you please add a call to `detect_sharp_edges()` function in the cpp example, so that we can test the code?
> 
> When it works, it would also be great if you could add another action/submenu to the demo, in the file `Polyhedron\demo\Polyhedron\Plugins\PMP\Detect_sharp_edges_plugin.cpp`
> It is the easiest way to visually test a new function and debug it

Ok. I added detect_sharp_edges to the example. I'm working on the code to add detect_sharp_corners to the plugin.

**Comment by janetournois** [comment]
sorry I meant `detect_sharp_corners()` of course!

**Comment by kabirkedia** [comment]
> sorry I meant `detect_sharp_corners()` of course!

That I've already added!!

**Comment by kabirkedia** [comment]
I have added the plugin. Kindly review it and let me know if any changes need to be made.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/detect_features.h [hunk 3]
```diff
@@ -476,6 +596,15 @@ void detect_sharp_edges(PolygonMesh& p,
   detect_sharp_edges(p, angle_in_deg, edge_is_feature_map,
                      parameters::all_default());
 }
+  
+template<typename PolygonMesh, typename VertexIsFeatureMap, typename FT>
+void detect_sharp_corners(PolygonMesh &p,
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/detect_features.h:601]: `const PolygonMesh& p` should be the last parameter (before NP)

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Detect_sharp_corners_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
+#include <QApplication>
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/Detect_sharp_corners_plugin.cpp:1]: please do not add another plugin, and simply enrich the existing `Detect_sharp_edges_plugin.cpp` as discussed by email

File: Polyhedron/demo/Polyhedron/Polyhedron_demo_detect_sharp_corners.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#pragma once
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Polyhedron_demo_detect_sharp_corners.h:1]: these functions should be moved to the `Detect_sharp_edges_plugin` files too, in the `Plugins/PMP` folder

**Review by janetournois — CHANGES_REQUESTED** [review]

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/detect_features.h [hunk 6]
```diff
@@ -66,6 +66,27 @@ is_sharp(PolygonMesh& polygonMesh,
...
+{
+  typedef typename boost::graph_traits<PolygonMesh>::edge_descriptor edge_descriptor;
+  if(is_border(edge(he,polygonMesh),polygonMesh)){
+    return false;
+  }
+  edge_descriptor e1 = edge(he,polygonMesh);
+  edge_descriptor e2 = edge(next(he,polygonMesh),polygonMesh);
+
+
+  if ( e1 * e2 <= cos_angle )
```
[the remark this answers is not in the document]
Comment (kabirkedia) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/detect_features.h:85]: What should I replace this with? In the sharp_edges example, the normal of the faces is computed to calculate the angle. Should I convert edges into a vector and then multiply them.

[the remark this answers is not in the document]
Comment (janetournois) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/detect_features.h:83]: You should compute an angle cosine, not a product of edge indices

**Pull request closed without merging**